src/material_library.cpp

```diff
@@ -0,0 +1,221 @@
+#include <unistd.h>
+#include <iostream>
+
+#ifndef PYNE_IS_AMALGAMATED
+#include "material_library.h"
+#endif
+
+namespace pyne {
+// Empty Constructor
+MaterialLibrary::MaterialLibrary(){};
+
+// Default constructor
+MaterialLibrary::MaterialLibrary(std::string file, std::string datapath) {
+  if (!check_file_exists(file)) {
+    std::cerr << "The file " << file << " does not exist or is read protected"
+              << std::endl;
+    exit(1);
+  }
+  if (!hdf5_path_exists(file, datapath)) {
+    std::cerr << "The datapath " << datapath << " in " << file << " empty."
+              << std::endl;
+    exit(1);
+  }
+
+  // load materials
+  from_hdf5(file);
+};
+
+// Destructor
+MaterialLibrary::~MaterialLibrary(){};
+
+// convert convert a filename into path+filename (for pyne)
+std::string MaterialLibrary::get_full_filepath(char* filename) {
+  std::string file(filename);
+  return MaterialLibrary::get_full_filepath(file);
+}
+
+// convert convert a filename into path+filename (for pyne)
+std::string MaterialLibrary::get_full_filepath(std::string filename) {
+  // remove all extra whitespace
+  filename.erase(std::remove(filename.begin(), filename.end(), ' '),
+                 filename.end());
+  // use stdlib call
+  const char* full_filepath = realpath(filename.c_str(), NULL);
+  return std::string(full_filepath);
+}
+
+// see if file exists
+bool MaterialLibrary::check_file_exists(std::string filename) {
```

Comment:
Same comment as above re: `const std::string& string`.

In general, these string arguments can move to a const reference to avoid memory copying.
@pshriwise

Reply:
make sense! I never think about them.